<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [antlr-interest] Development of an XQuery parser with	full-text	extensions, project report
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:antlr-interest%40antlr.org?Subject=Re:%20%5Bantlr-interest%5D%20Development%20of%20an%20XQuery%20parser%20with%0A%09full-text%09extensions%2C%20project%20report&In-Reply-To=%3C1198762334.4773a95e4e85d%40clients.latnet.lv%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="025356.html">
   <LINK REL="Next"  HREF="025362.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[antlr-interest] Development of an XQuery parser with	full-text	extensions, project report</H1>
    <B>Guntis Ozols</B> 
    <A HREF="mailto:antlr-interest%40antlr.org?Subject=Re:%20%5Bantlr-interest%5D%20Development%20of%20an%20XQuery%20parser%20with%0A%09full-text%09extensions%2C%20project%20report&In-Reply-To=%3C1198762334.4773a95e4e85d%40clients.latnet.lv%3E"
       TITLE="[antlr-interest] Development of an XQuery parser with	full-text	extensions, project report">guntiso at latnet.lv
       </A><BR>
    <I>Thu Dec 27 05:32:14 PST 2007</I>
    <P><UL>
        <LI>Previous message: <A HREF="025356.html">[antlr-interest] Development of an XQuery parser with full-text extensions, project report
</A></li>
        <LI>Next message: <A HREF="025362.html">[antlr-interest] Development of an XQuery parser with full-text	extensions, project report
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#25361">[ date ]</a>
              <a href="thread.html#25361">[ thread ]</a>
              <a href="subject.html#25361">[ subject ]</a>
              <a href="author.html#25361">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>&gt;<i> &gt; nextToken() catches RuleMismatchException (could be another kindof
</I>&gt;<i> &gt; exception), so it doesn't declare that it throws
</I>&gt;<i> &gt; RuleMismatchException.
</I>&gt;<i> &gt; But there are instances, where someone wants overwrite nextToken()
</I>&gt;<i> &gt; with rethrowing the exception - and can't.
</I>&gt;<i>
</I>&gt;<i> Hmm...well, I just looked and it seems like it will cause a ripple
</I>&gt;<i> effect where you have to put the catch clause everywhere.
</I>
Why catch? You only need to catch lexer/parser predicates,
but predicates are normal flow, and exceptions are slow
so I'd suggest to call special (with flag?) methods for
predicates to maybe return null instead.

&gt;<i> It should be okay as most of the references are within a parser rule,
</I>&gt;<i> but don't you think it is simply easier to throw either an error
</I>
No, errors are intended for more serious problems,
like OutOfMemoryError, StackOverflowError

&gt;<i> or run-time exception if you want the entire parsing process
</I>&gt;<i> to stop upon lexical error.
</I>
This must be on by default.
BTW LexerException can extend RuntimeException, too,
to make throws declaration optional.

&gt;<i> Throwing a recognition exception from a token stream is undefined.
</I>&gt;<i> Who is supposed to catch it? you might not even use a parser?
</I>
main() or whatever is supposed to catch or not to catch it,
depending on circumstances which are known to lexer/parser user,
but not known to lexer/parser developer.

&gt;<i> I'm not sure recognition exceptions should come out.
</I>&gt;<i> When there is a problem it should keep going.
</I>
Oh NO! Reminds me:
  expecting zip-file, getting atomic-bomb, exploding anyway...

&gt;<i> If you wanted to bail out and stop lexing, you
</I>&gt;<i> must make the invoking parser fail as well.
</I>
Sure

&gt;<i> I don't think we should pass a lexical recognition exception
</I>&gt;<i> on to the parser because it is not an exception for the parser.
</I>
It is not for the parser, and therefore should be PASSED TO and
NOT HANDLED by the parser, UNLESS rulecatch LexerException is
defined in parser grammar.

By default, if LexerException extends RuntimeException,
it may even be not mentioned in parser, still cause parse to fail.

&gt;<i>    Does that make sense?
</I>
See above :)

&gt;<i> Ter
</I>
Guntis

</PRE>



<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="025356.html">[antlr-interest] Development of an XQuery parser with full-text extensions, project report
</A></li>
	<LI>Next message: <A HREF="025362.html">[antlr-interest] Development of an XQuery parser with full-text	extensions, project report
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#25361">[ date ]</a>
              <a href="thread.html#25361">[ thread ]</a>
              <a href="subject.html#25361">[ subject ]</a>
              <a href="author.html#25361">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://www.antlr.org/mailman/listinfo/antlr-interest">More information about the antlr-interest
mailing list</a><br>
</body></html>
